File: JUNE 23/06 June/Predecessor and Successor gfg.cpp
class Solution
{
    public:
    void traverse(Node* root, Node* &pre, Node* &suf, int preVal, int sufVal, int key)
    {
        if(!root) return;
        int val = root->key;
        if(val>key && val<sufVal)
        {
            sufVal = val;
            suf = root;
        }
        if(val<key && val>preVal)
        {
            preVal = val;
            pre = root;
        }
        traverse(root->left, pre, suf, preVal, sufVal, key);
        traverse(root->right, pre, suf, preVal, sufVal, key);
        
    }
    void findPreSuc(Node* root, Node*& pre, Node*& suf, int key)
    {
        // Your code goes here
        int sufVal=INT_MAX, preVal=INT_MIN;
        traverse(root, pre, suf, preVal, sufVal, key);
    }
};

File: MAY 2025/19 May/Predecessor and Successor gfg.java
class Solution {
    private int predecessor = -1, successor = -1;
    private void traverse(Node root, ArrayList<Node> traversal)
    {
        if(root == null)
            return;
        traverse(root.left, traversal);
        traversal.add(root);
        traverse(root.right, traversal);
    }
    public ArrayList<Node> findPreSuc(Node root, int key) {
        // code here
        ArrayList<Node> traversal = new ArrayList<>();
        traverse(root, traversal);
        ArrayList<Node> list = new ArrayList<>();
        list.add(null);
        list.add(null);
        int N = traversal.size();
        for(int i=0; i<N; i++)
        {
            if(traversal.get(i).data<key)
                list.set(0,traversal.get(i));
                
            if(list.get(1)==null && traversal.get(i).data>key)
                list.set(1, traversal.get(i));
        }
        return list;
    }
}
